pin/set.go

```diff
@@ -0,0 +1,354 @@
+}
+
+type itemIterator func() (k util.Key, data []byte, ok bool)
+
+type keyObserver func(util.Key)
+
+// refcount is the marshaled format of refcounts. It may change
+// between versions; this is valid for version 1. Changing it may
+// become desirable if there are many links with refcount > 255.
+//
+// There are two guarantees that need to be preserved, if this is
+// changed:
+//
+//     - the marshaled format is of fixed size, matching
+//       unsafe.Sizeof(refcount(0))
+//     - methods of refcount handle endianness, and may
+//       in later versions need encoding/binary.
+type refcount uint8
+
+func (r refcount) Bytes() []byte {
+	return []byte{byte(r)}
+}
+
+// readRefcount returns the idx'th refcount in []byte, which is
+// assumed to be a sequence of refcount.Bytes results.
+func (r *refcount) ReadFromIdx(buf []byte, idx int) {
+	*r = refcount(buf[idx])
+}
+
+type sortByHash struct {
+	links []*merkledag.Link
+	data  []byte
+}
+
+func (s sortByHash) Len() int {
+	return len(s.links)
+}
+
+func (s sortByHash) Less(a, b int) bool {
+	return bytes.Compare(s.links[a].Hash, s.links[b].Hash) == -1
+}
+
+func (s sortByHash) Swap(a, b int) {
+	s.links[a], s.links[b] = s.links[b], s.links[a]
+	if len(s.data) != 0 {
+		const n = int(unsafe.Sizeof(refcount(0)))
+		tmp := make([]byte, n)
+		copy(tmp, s.data[a*n:a*n+n])
+		copy(s.data[a*n:a*n+n], s.data[b*n:b*n+n])
+		copy(s.data[b*n:b*n+n], tmp)
+	}
+}
+
+func storeItems(ctx context.Context, dag merkledag.DAGService, estimatedLen uint64, iter itemIterator, internalKeys keyObserver) (*merkledag.Node, error) {
+	seed, err := randomSeed()
+	if err != nil {
+		return nil, err
+	}
+	n := &merkledag.Node{
+		Links: make([]*merkledag.Link, 0, defaultFanout+maxItems),
+	}
+	for i := 0; i < defaultFanout; i++ {
+		n.Links = append(n.Links, &merkledag.Link{Hash: emptyKey.ToMultihash()})
+	}
+	internalKeys(emptyKey)
+	hdr := &pb.Set{
+		Version: proto.Uint32(1),
+		Fanout:  proto.Uint32(defaultFanout),
+		Seed:    proto.Uint32(seed),
+	}
+	if err := writeHdr(n, hdr); err != nil {
+		return nil, err
+	}
+	hdrLen := len(n.Data)
+
+	if estimatedLen < maxItems {
+		// it'll probably fit
+		for i := 0; i < maxItems; i++ {
+			k, data, ok := iter()
+			if !ok {
+				// all done
+				break
+			}
+			n.Links = append(n.Links, &merkledag.Link{Hash: k.ToMultihash()})
+			n.Data = append(n.Data, data...)
+		}
+		// sort by hash, also swap item Data
+		s := sortByHash{
+			links: n.Links[defaultFanout:],
+			data:  n.Data[hdrLen:],
+		}
+		sort.Stable(s)
+	}
+
+	// wasteful but simple
+	type item struct {
+		k    util.Key
+		data []byte
+	}
+	hashed := make(map[uint32][]item)
+	for {
+		k, data, ok := iter()
+		if !ok {
+			break
+		}
+		h := hash(seed, k)
+		hashed[h] = append(hashed[h], item{k, data})
+	}
+	for h, items := range hashed {
+		childIter := func() (k util.Key, data []byte, ok bool) {
+			if len(items) == 0 {
+				return "", nil, false
+			}
+			first := items[0]
+			items = items[1:]
```

Comment:
overwriting `items` here is pretty sketch-- i would be worried about editing this function. another way? or comments?